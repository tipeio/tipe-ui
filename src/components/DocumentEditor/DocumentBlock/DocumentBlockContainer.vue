<template>
  <tipe-document-block
    :type="type"
    :value="value"
    :name="name"
    :status="status"
    :success-message="successMessage"
    :warning-message="warningMessage"
    :error-message="errorMessage"
  >
    <component
      :status="status"
      :value="value"
      :size="large"
      :is="is"
    />
  </tipe-document-block>
</template>

<script>
import documentBlockShape from '@/types/DocumentBlock'
import TipeDocumentBlock from './DocumentBlock'
import TipeSwitch from '@/components/Switch'

export default {
  name: 'DocumentBlockContainer',
  components: {
    TipeDocumentBlock,
    TipeSwitch
  },
  props: documentBlockShape,
  computed: {
    is() {
      return 'tipe-switch'
    }
  }
}
</script>
