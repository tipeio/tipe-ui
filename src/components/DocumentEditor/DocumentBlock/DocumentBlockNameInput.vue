<template>
  <input
    :data-tipe-ui="$options.name"
    :value="value"
    :disabled="disabled || waiting"
    :class="rootClassObject"
    name="name"
    type="text"
    placeholder="block name"
    @change="$emit('change', $event)"
  >
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeDocumentBlockNameInput',
  props: {
    value: vueTypes.string,
    waiting: vueTypes.bool.def(false),
    disabled: vueTypes.bool.def(false)
  },
  computed: {
    rootClassObject() {
      return {
        waiting: this.waiting
      }
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentBlockNameInput'] {
  border: none;
  outline: none;
  margin: 0;
  padding: 0;
  background: transparent;
  height: 100%;
  width: 100%;
  display: flex;
  align-content: center;
  justify-content: flex-start;
}

[data-tipe-ui='TipeDocumentBlockNameInput']:disabled {
  cursor: not-allowed;
}

[data-tipe-ui='TipeDocumentBlockNameInput'].waiting {
  cursor: wait;
}
</style>
