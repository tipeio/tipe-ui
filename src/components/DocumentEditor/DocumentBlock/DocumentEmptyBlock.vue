<template>
  <div
    :data-tipe-ui="$options.name"
  >
    <div class="body">
      Please add a block.
    </div>
  </div>
</template>

<script>
export default {
  name: 'TipeDocumentEmptyBlock'
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentEmptyBlock'] {
  display: flex;
  flex-direction: column;
  border-radius: 0.1875rem;
  box-sizing: border-box;
  width: 100%;
  min-width: 20rem;
}

.margin {
  grid-area: margin;
}

.body {
  grid-area: body;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  color: var(--midPurple);
}
</style>
