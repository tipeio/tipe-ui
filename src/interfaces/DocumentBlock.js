import vueTypes from 'vue-types'
import DocumentBlockValidation from './DocumentBlockValidation'

export default {
  id: vueTypes.string.def(''),
  type: vueTypes.string.def(''),
  label: vueTypes.string.def(''),
  name: vueTypes.string.def(''),
  apiId: vueTypes.string.def(''),
  value: vueTypes.any,
  description: vueTypes.string.def(''),
  status: vueTypes.oneOf(['success', 'error', 'warning', '']),
  successMessage: vueTypes.string.def(''),
  errorMessage: vueTypes.string.def(''),
  warningMessage: vueTypes.string.def(''),
  waiting: vueTypes.bool.def(false),
  disabled: vueTypes.bool.def(false),
  validation: vueTypes.shape(DocumentBlockValidation)
}
