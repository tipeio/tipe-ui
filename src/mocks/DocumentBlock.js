import faker from 'faker'
import { getBlockTypes } from '@/libs/blocks'
import documentBlockValidationMock from '@/mocks/DocumentBlockValidation'

export default documentBlock => ({
  type: getBlockTypes()[
    faker.random.number({ min: 0, max: getBlockTypes().length - 1 })
  ],
  value: '',
  name: faker.lorem.word(),
  status: ['success', 'warning', 'error', ''][
    faker.random.number({ min: 0, max: 2 })
  ],
  successMessage: faker.lorem.sentence(),
  errorMessage: faker.lorem.sentence(),
  warningMessage: faker.lorem.sentence(),
  waiting: faker.random.boolean(),
  disabled: faker.random.boolean(),
  validation: documentBlockValidationMock(),
  ...documentBlock
})
