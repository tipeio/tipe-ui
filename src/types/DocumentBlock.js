import vueTypes from 'vue-types'
import DocumentBlockValidation from '@/types/DocumentBlockValidation'
import blockTypes from '@/types/BlockTypes'

export default {
  type: vueTypes.oneOf(blockTypes).isRequired,
  name: vueTypes.string,
  value: vueTypes.any,
  status: vueTypes.oneOf(['success', 'error', 'warning', '']),
  successMessage: vueTypes.string.def(''),
  errorMessage: vueTypes.string.def(''),
  warningMessage: vueTypes.string.def(''),
  waiting: vueTypes.bool.def(false),
  disabled: vueTypes.bool.def(false),
  validation: vueTypes.shape(DocumentBlockValidation)
}
