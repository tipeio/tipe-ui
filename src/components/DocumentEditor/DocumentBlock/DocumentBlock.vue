<template>
  <div
    :data-tipe-ui="$options.name"
    :class="rootStyleObject"
    @mouseover="hover = true"
    @mouseout="hover = false"
  >
    <div class="header">
      <tipe-document-block-header
        v-bind="block"
        @change="value => $emit('change', value)"
      />
    </div>
    <div class="body">
      <tipe-document-block-value-input
        :status="block.status"
        :value="block.value"
        :waiting="block.waiting"
        :disabled="block.disabled"
        :type="block.type"
        @change="value => $emit('change', value)"
      />
    </div>
    <div class="footer">
      <tipe-document-block-message
        v-if="block.status != ''"
        :success-message="block.errorMessage"
        :warning-message="block.errorMessage"
        :error-message="block.errorMessage"
        :status="block.status"
      />
    </div>
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import interfaces from '@tipe/tipe-interfaces'
import TipeDocumentBlockValueInput from './DocumentBlockValueInput'
import TipeDocumentBlockHeader from './DocumentBlockHeader'
import TipeDocumentBlockMessage from './DocumentBlockMessage'

export default {
  name: 'TipeDocumentBlock',
  components: {
    TipeDocumentBlockValueInput,
    TipeDocumentBlockHeader,
    TipeDocumentBlockMessage
  },
  props: {
    block: vueTypes.shape(interfaces.documentBlock)
  },
  data() {
    return {
      hover: false
    }
  },
  computed: {
    rootStyleObject() {
      return {
        [`status-${this.block.status}`]: this.block.status,
        waiting: this.block.waiting,
        disabled: this.block.disabled
      }
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentBlock'] {
  display: flex;
  flex-direction: column;
  border-radius: 0.1875rem;
  box-sizing: border-box;
  width: 100%;
  min-width: 20rem;
}

.margin {
  grid-area: margin;
}

.header {
  grid-area: header;
}

.body {
  grid-area: body;
}

.footer {
  grid-area: footer;
}

[data-tipe-ui='TipeDocumentBlock'].waiting {
  cursor: wait;
}

[data-tipe-ui='TipeDocumentBlock'].disabled {
  cursor: not-allowed;
}
</style>
