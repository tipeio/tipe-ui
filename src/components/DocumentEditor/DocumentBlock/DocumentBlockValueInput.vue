<template>
  <div :data-tipe-ui="$options.name">
    <component
      :status="status"
      :value="value"
      :waiting="waiting"
      :disabled="disabled"
      :is="is"
      size="medium"
      @change="onChange"
    />
  </div>
</template>

<script>
import interfaces from '@/interfaces'
import { getBlockValueComponent } from '../helpers'
import TipeSwitch from '@/components/Switch'
import TipeNumberInput from '@/components/NumberInput'
import TipeTextInput from '@/components/TextInput'

export default {
  name: 'TipeDocumentBlock',
  components: {
    TipeSwitch: TipeSwitch,
    TipeNumberInput: TipeNumberInput,
    TipeTextInput: TipeTextInput
  },
  props: interfaces.documentBlock,
  computed: {
    is() {
      return getBlockValueComponent(this.type)
    }
  },
  methods: {
    onChange(value) {
      this.$emit('change', {
        name: this.name,
        value
      })
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeDocumentBlock'] {
  padding: 1.5rem 0 1rem 0;
}
</style>
