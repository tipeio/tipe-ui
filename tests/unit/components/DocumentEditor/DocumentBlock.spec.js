import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import DocumentBlock from '@/components/DocumentEditor/DocumentBlock'
import DocumentBlockNameInput from '@/components/DocumentEditor/DocumentBlock/DocumentBlockNameInput'
import DocumentBlockValueInputStub from '../../stubs/DocumentBlockValueInputStub'
import documentBlockFixture from '@/fixtures/DocumentBlock'
import documentBlockMock from '@/mocks/DocumentBlock'
import faker from 'faker'

describe('DocumentBlock', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const propsData = documentBlockFixture()
      const wrapper = shallowMount(DocumentBlock, { propsData })

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('has correct data-tipe-iu attibute', () => {
      const propsData = documentBlockMock()
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeDocumentBlock')
    })

    it('contains a DocumentBlockNameInput component', () => {
      const propsData = documentBlockMock()
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.contains(DocumentBlockNameInput)).toEqual(true)
    })
  })

  describe(':props', () => {
    it(':name - should pass name to DocumentBlockNameInput component value prop', () => {
      const propsData = documentBlockMock()

      const wrapper = shallowMount(DocumentBlock, { propsData })
      expect(wrapper.find(DocumentBlockNameInput).props().value).toEqual(
        propsData.name
      )
    })

    it(':status(undefined) - no status class ', () => {
      const propsData = documentBlockMock({ status: '' })
      const wrapper = shallowMount(DocumentBlock, { propsData })
      expect(wrapper.classes()).not.toEqual(
        expect.arrayContaining([expect.stringMatching(/status-/)])
      )
    })

    it(':status(success) - has status-success class', () => {
      const propsData = documentBlockMock({ status: 'success' })
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('status-success')
    })

    it(':status(error) - has status-error class', () => {
      const propsData = documentBlockMock({ status: 'error' })
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('status-error')
    })

    it(':status(warning) - has status-warning class', () => {
      const propsData = documentBlockMock({ status: 'error' })
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('status-error')
    })

    it(':waiting - should pass waiting to DocumentBlockNameInput component', () => {
      const propsData = documentBlockMock({ waiting: true })
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.find(DocumentBlockNameInput).props().waiting).toEqual(true)
    })

    it(':waiting(false) - no waiting class', () => {
      const propsData = documentBlockMock({ waiting: false })
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).not.toEqual(expect.arrayContaining(['waiting']))
    })

    it(':waiting(true) - has waiting class', () => {
      const propsData = documentBlockMock({ waiting: true })
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.classes()).toContain('waiting')
    })

    it(':disabled - should pass disabled to DocumentBlockNameInput component', () => {
      const propsData = documentBlockMock({ disabled: true })
      const wrapper = shallowMount(DocumentBlock, { propsData })

      expect(wrapper.find(DocumentBlockNameInput).props().disabled).toEqual(
        true
      )
    })
  })

  describe('@events', () => {
    it('@change - should emit change when name changed', () => {
      const propsData = documentBlockMock({ disabled: false, waiting: false })
      const stubs = { TipeDocumentBlockNameInput: DocumentBlockNameInput }
      const wrapper = shallowMount(DocumentBlock, { propsData, stubs })

      wrapper.find(DocumentBlockNameInput).trigger('change')

      expect(wrapper.emitted().change).toBeTruthy()
      expect(wrapper.emitted().change[0][0]).toMatchObject({
        name: propsData.name
      })
    })

    it('@change - should emit change when value changed', () => {
      const propsData = documentBlockMock({
        disabled: false,
        waiting: false,
        value: faker.lorem.word()
      })
      const stubs = {
        'tipe-document-block-value-input': DocumentBlockValueInputStub
      }
      const wrapper = shallowMount(DocumentBlock, { propsData, stubs })

      wrapper.find(DocumentBlockValueInputStub).trigger('change')
      expect(wrapper.emitted().change).toBeTruthy()
      expect(wrapper.emitted().change[0][0]).toMatchObject({
        value: propsData.value
      })
    })
  })
})
