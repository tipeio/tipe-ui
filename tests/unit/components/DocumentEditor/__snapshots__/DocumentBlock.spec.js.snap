// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentBlock <template> matches previous snapshot 1`] = `
<div data-tipe-ui="TipeDocumentBlock" class="status-success">
  <div class="margin">
    <!---->
  </div>
  <div class="header">
    <TipeDocumentBlockHeader-stub type="SIMPLE_TEXT" label="Simple Text" name="test-name" apiId="test-apiId" value="test-value" description="test-description" status="success" successMessage="" errorMessage="" warningMessage="" validation="[object Object]"></TipeDocumentBlockHeader-stub>
  </div>
  <div class="body">
    <TipeDocumentBlockValueInput-stub type="SIMPLE_TEXT" label="" name="" apiId="" value="test-value" description="" status="success" successMessage="" errorMessage="" warningMessage=""></TipeDocumentBlockValueInput-stub>
  </div>
  <div class="footer">
    <TipeDocumentBlockMessage-stub status="success" successMessage="" warningMessage="" errorMessage=""></TipeDocumentBlockMessage-stub>
  </div>
</div>
`;
